dawn/src/dawn/CodeGen/Cuda-ico/CudaIcoCodeGen.cpp

```diff
@@ -589,14 +633,21 @@ void CudaIcoCodeGen::generateAllCudaKernels(
         break;
       }
 
-      // k loop
-      cudaKernel.addBlockStatement("for(int kIter = klo; kIter < khi; kIter++)", [&]() {
-        cudaKernel.addBlockStatement("if (kIter >= kSize)",
-                                     [&]() { cudaKernel.addStatement("return"); });
+      std::stringstream k_size;
+      if(interval.levelIsEnd(iir::Interval::Bound::upper)) {
+        k_size << "kSize + " << interval.upperOffset();
+      } else {
+        k_size << interval.upperLevel() << " + " << interval.upperOffset();
+      }
 
+      // k loop
+      for(const auto& doMethodPtr : stage->getChildren()) {
         // Generate Do-Method
-        for(const auto& doMethodPtr : stage->getChildren()) {
-          const iir::DoMethod& doMethod = *doMethodPtr;
+        const iir::DoMethod& doMethod = *doMethodPtr;
+        cudaKernel.addBlockStatement("for(int kIter = klo; kIter < khi; kIter++)", [&]() {
+          cudaKernel.addBlockStatement("if (kIter >= " + k_size.str() + ")",
+                                       [&]() { cudaKernel.addStatement("return"); });
```

Comment:
If all do methods in a stage are on the same k-interval, the iteration over k (and the check that it doesn't exceed the upper bound) can be done outside of the loop over do methods, correct?